src/local_scheduler/local_scheduler_algorithm.cc

```diff
@@ -1198,41 +1201,42 @@ void handle_object_removed(LocalSchedulerState *state,
+  delete entry->dependent_tasks;
   free(entry);
 
   /* Track queued tasks that were dependent on this object.
    * NOTE: Since objects often get removed in batches (e.g., during eviction),
    * we may end up iterating through the queues many times in a row. If this
    * turns out to be a bottleneck, consider tracking dependencies even for
    * tasks in the dispatch queue, or batching object notifications. */
-  task_queue_entry *elt, *tmp;
-  /* Track the dependency for tasks that were in the waiting queue. */
-  DL_FOREACH(algorithm_state->waiting_task_queue, elt) {
-    TaskSpec *task = elt->spec;
-    int64_t num_args = TaskSpec_num_args(task);
-    for (int i = 0; i < num_args; ++i) {
-      if (TaskSpec_arg_by_ref(task, i)) {
-        ObjectID arg_id = TaskSpec_arg_id(task, i);
-        if (ObjectID_equal(arg_id, removed_object_id)) {
-          fetch_missing_dependency(state, algorithm_state, elt,
-                                   removed_object_id);
-        }
-      }
-    }
-  }
   /* Track the dependency for tasks that were in the dispatch queue. Remove
    * these tasks from the dispatch queue and push them to the waiting queue. */
-  DL_FOREACH_SAFE(algorithm_state->dispatch_task_queue, elt, tmp) {
-    TaskSpec *task = elt->spec;
-    int64_t num_args = TaskSpec_num_args(task);
+  for (auto it = algorithm_state->dispatch_task_queue->begin();
+       it != algorithm_state->dispatch_task_queue->end();) {
+    TaskQueueEntry task = *it;
+    if (TaskSpec_is_dependent_on(task.spec, removed_object_id)) {
+      /* This task was dependent on the removed object. */
+      LOG_DEBUG("Moved task from dispatch queue back to waiting queue");
+      algorithm_state->waiting_task_queue->push_back(task);
+      /* Remove the task from the dispatch queue, but do not free the task
+       * spec. */
+      it = algorithm_state->dispatch_task_queue->erase(it);
+    } else {
+      /* The task can still run,so continue to the next task. */
+      ++it;
+    }
+  }
+
+  /* Track the dependency for tasks that are in the waiting queue, including
+   * those that were just moved from the dispatch queue. */
+  for (auto it = algorithm_state->waiting_task_queue->begin();
+       it != algorithm_state->waiting_task_queue->end(); ++it) {
+    int64_t num_args = TaskSpec_num_args(it->spec);
     for (int i = 0; i < num_args; ++i) {
-      if (TaskSpec_arg_by_ref(task, i)) {
-        ObjectID arg_id = TaskSpec_arg_id(task, i);
+      if (TaskSpec_arg_by_ref(it->spec, i)) {
+        ObjectID arg_id = TaskSpec_arg_id(it->spec, i);
         if (ObjectID_equal(arg_id, removed_object_id)) {
-          LOG_DEBUG("Moved task from dispatch queue back to waiting queue");
-          DL_DELETE(algorithm_state->dispatch_task_queue, elt);
-          DL_APPEND(algorithm_state->waiting_task_queue, elt);
-          fetch_missing_dependency(state, algorithm_state, elt,
+          fetch_missing_dependency(state, algorithm_state, it,
```

Comment:
Yes, we want to do that, since multiple tasks may depend on the same object ID. The same task will never appear more than once under the same object ID though.